Added cEUR mainnet proposals (#192)
celo-org/celo-proposals: CGPs/0023.md

> This governance proposal enables oracle addresses to make them able to report the value of CELO/EUR for the cEUR contract, deployed on [CR3](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0022.md). It also proposed to send 100 CELO to each of the oracles to cover gas fees for the first couple of years and freezing StableTokenEUR (cEUR) and ExchangeEUR.

Reviewer: I think `It also proposed` -> `It also proposes`

Reviewer: Not super passionate but it may be good to rationalize the 100 CELO amount, eg how much we've seen existing oracles use and why 100 makes sense

> This governance proposal enables oracle addresses to make them able to report the value of CELO/EUR for the cEUR contract, deployed on [CR3](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0022.md). It also proposed to send 100 CELO to each of the oracles to cover gas fees for the first couple of years and freezing StableTokenEUR (cEUR) and ExchangeEUR.

Reviewer 2: ```suggestion
This governance proposal enables oracle addresses to make them able to report the value of CELO/EUR for the cEUR contract, deployed on [CR3](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0022.md). It also proposes to send 100 CELO to each of the oracles to cover gas fees for the first couple of years and freezing StableTokenEUR (cEUR) and ExchangeEUR.
```

> Following [CR3](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0022.md), cEUR will be in an status where it would be imposible to mint or burn due empty buckets that are not possible to update (because Mento buckets are zero and there are no oracles enable to updated), but cEUR will be possible to move. Freezing is meant to prevent cEUR from moving before enabling the stability mechanism.

Reviewer 2: ```suggestion
Following [CR3](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0022.md), cEUR will be in a status where it will be imposible to mint or burn due to empty buckets that are not possible to update (because Mento buckets are zero and there are no oracles enabled to update them), but cEUR will be transferable. Freezing is meant to prevent cEUR being transferred before enabling the stability mechanism.
```

> cEUR in circulation with no stability mechanism in place can lead to a depledge. Although the risk of cEUR moving before its full activation it's very low, this proposal will add an extra layer of security. The only two addresses holding cEUR after CR3 is deployed will by the [Community Fund](https://celocommunityfund.org/) (Governance Proxy) and a Celo Foundation wallet.

Reviewer 3: nit: I might be wrong but I think the correct term is `depeg` / `de-peg`. Depledge is something a bit different from my understanding:

> depeg
> To remove a previously instituted peg on a currency. For example, if Currency A is pegged to Currency B at a 1:1 ratio, but the central bank for Currency A decides to let it float, it is said to be depegged from Currency B. Depegging may occur if the peg is causing inflation or if the central bank is unable to sustain the peg for other reasons.

---

> depledge
> (transitive, intransitive) To revoke one's pledge of allegiance to a college fraternity.

> cEUR in circulation with no stability mechanism in place can lead to a depledge. Although the risk of cEUR moving before its full activation it's very low, this proposal will add an extra layer of security. The only two addresses holding cEUR after CR3 is deployed will by the [Community Fund](https://celocommunityfund.org/) (Governance Proxy) and a Celo Foundation wallet.

Reviewer: `CR3 is deployed will by` -> `CR3 is deployed will be`

> cEUR in circulation with no stability mechanism in place can lead to a depledge. Although the risk of cEUR moving before its full activation it's very low, this proposal will add an extra layer of security. The only two addresses holding cEUR after CR3 is deployed will by the [Community Fund](https://celocommunityfund.org/) (Governance Proxy) and a Celo Foundation wallet.

Author: ```suggestion
cEUR in circulation with no stability mechanism in place can lead to a de-peg. Although the risk of cEUR moving before its full activation it's very low, this proposal will add an extra layer of security. The only two addresses holding cEUR after CR3 is deployed will be the [Community Fund](https://celocommunityfund.org/) (Governance Proxy) and a Celo Foundation wallet.
```

Please note this proposal shall be submited after CR3 is proposed, because it needs an address for StableTokenEUR, but it can be executed before or after.
 
[CGP #26](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0022.md) was merged into this proposal after Baklava deployment.

## Proposed Changes

> 1. The first ten transactions enable a specific address to report to StableTokenEUR

Reviewer: nit: make it clear that this is enabling multiple addresses to report to SortedOracles for the StableTokenEUR rate

>   - Destination: SortedOragles.addOracle

Reviewer: SortedOragles -> SortedOracles

Author: Thanks! 🥴

  - Value: 0
2. The following ten transactions are to fund the oracles:
  - Destination: GoldToken.transfer
  - Data: first argument is the StableTokenEUR address and the second one the address of the oracle to be added, same as in (1)
  - Value: 100e18  (note the value field of the proposal it's actually `0`, as this was implemented as a token transfer)
3. Freezing StableTokenEUR
  - Destination: Freezer.freeze
  - Data: StableTokenEUR (`0xD8763CBa276a3738E6DE85b4b3bF5FDed6D6cA73`)
  - Value: 0
4. Freezing ExchangeEUR
  - Destination: Freezer.freeze
  - Data: ExchangeEUR (`0xE383394B913d7302c49F794C7d3243c429d53D1d`)
  - Value: 0
## Verification

Can be left as “TODO” until the proposal is made.

## Risks

- Adding new oracles will add load to the network, which is not a problem at this stage but could become one as new stable assets and other projects deploy new oracles and network adoption grows.

> - Oracle rate risk: In case of imprecise CELO/USD oracle rates and/or manipulated CELO/USD market rates, users could exchange cUSD or CELO with the reserve at a price that does not reflect the current market valuation. This could negatively affect the reserve collateralization ratio and therefore long-term cUSD stability in two ways: First, a user could mint X cUSD while providing less than X USD worth of reserve collateral and second, a user could redeem Y cUSD for more than Y USD worth of reserve collateral. The major mitigant of this risk is the initial reduction of the reserveFraction parameter as explained in the overview section which comes at the cost of the risks described further below. Additional mitigants include oracle client features like a circuit breaker protecting against unreasonably large price movements. (Same as [CGP #0007](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0007.md))

Reviewer 3: Should this reference CELO/EUR cEUR, or if not make it clear why we're referencing that here, is it because they share the reserver?

Author: True!

> - Oracle rate risk: In case of imprecise CELO/USD oracle rates and/or manipulated CELO/USD market rates, users could exchange cUSD or CELO with the reserve at a price that does not reflect the current market valuation. This could negatively affect the reserve collateralization ratio and therefore long-term cUSD stability in two ways: First, a user could mint X cUSD while providing less than X USD worth of reserve collateral and second, a user could redeem Y cUSD for more than Y USD worth of reserve collateral. The major mitigant of this risk is the initial reduction of the reserveFraction parameter as explained in the overview section which comes at the cost of the risks described further below. Additional mitigants include oracle client features like a circuit breaker protecting against unreasonably large price movements. (Same as [CGP #0007](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0007.md))

Author: ```suggestion
- Oracle rate risk: In case of imprecise CELO/EUR oracle rates and/or manipulated CELO/EUR market rates, users could exchange cEUR or CELO with the reserve at a price that does not reflect the current market valuation. This could negatively affect the reserve collateralization ratio and therefore long-term cEUR stability in two ways: First, a user could mint X cEUR while providing less than X EUR worth of reserve collateral and second, a user could redeem Y cEUR for more than Y EUR worth of reserve collateral. The major mitigant of this risk is the initial reduction of the reserveFraction parameter as explained in the overview section which comes at the cost of the risks described further below. Additional mitigants include oracle client features like a circuit breaker protecting against unreasonably large price movements. (Same as [CGP #0007](https://github.com/celo-org/celo-proposals/blob/master/CGPs/0007.md))
```
